### ISSUE TITLE
multiple homebridge instances

### ISSUE THREAD

**Issue by zork17** [nfarina/homebridge#1160]
I'm trying to run 2 homebridge instances to overcome the 100 accessories limit.
The first instance has 90+ knx accessories.
The second homebridge instance points to a new location with a new config.json file, but when it launches it is looking for the knx_config.json file although there is no knx platform in the second instance.
When I put an empty knx_config.json file or one with different ip port I get an error:
Error: Cannot reach knxd or eibd service, please check installation and knx_config.json

Is there a way to tell homebridge not to load knx in an instance?

**Comment by snowdd1** [comment]
@nfarina : is there a way to stop an instance to load globally installed dynamic platforms? The user path is additionally only as I understood.

**Comment by KhaosT** [comment]
Currently no... As a workaround, maybe have user put a special flag in the configuration file for knx plugin to prevent it from doing discovery?

**Comment by zork17** [comment]
@KhaosT @snowdd1 
This is a good idea, since I already have a different knx_config file for each knx.
The first one has 90+ accessories and the second one is empty.

I also think it is a good idea generally, because in the documentation for almost every plugin (including knx) it written to install like this:
_sudo npm install -g homebridge-knx_
Many plugins that I installed, just put out a message saying the plugin loaded but didn't find a platfrom or accessories in the config.json and that's it. It doesn't affect the running of homebridge.
So maybe it could also be the case for knx: If there is no mention of knx platform in the config,json, it should not load?

**Comment by snowdd1** [comment]
@zork17 : I know. I used to recommend installing everything locally, but users got confused with the various installation options - it says globally on homebridge repo, did say locally at mine... more trouble than use, so I changed that to what seemed common agreement to me. However, it does not make things easier, as we see here, and even `homebridge` isn't a command that you would use in daily life. It's a server. You put it in a start configuration and forget about it.  

However, I could change the "bootloader" of homebridge-knx. @KhaosT : The plugin loader requires the platform's main function, providing the API, but no config section yet - So I cannot use a config.json entry there. I could check for a user file there (like `./homebridge/knx-ignore.txt`) and prevent `homebridgeAPI.registerPlatform("homebridge-knx", "KNX", KNXPlatform, true);` from being called. That would be sufficient, wouldn't it?

**Comment by snowdd1** [comment]
I have now created a branch `ignore-option` which contains the following in the main loader: 
```JavaScript
	/*
	 * Experimental: Look for a user file called knx-ignore.txt in the user config path.
	 * If it is there, exit here and DO NOT REGISTER the platform
	 */
	let fs = require('fs');
	let path = require('path');
	let checkfilepath = path.join(homebridgeAPI.user.storagePath(), 'knx-ignore.txt');
	if (fs.existsSync(checkfilepath)) {
		console.log('[WARNING] Found blocking file, exiting now. To load homebridge-knx, remove '+ checkfilepath);
		return;
	}
	// END OF INSERTION FOR BRANCH ignore-option
```

If the block file exists in the directory of config.json, it will refuse to start. My test looked like this:
```
[2017-02-24 10:16:55] Loaded plugin: homebridge-knx
homebridge API version: 2.1
[WARNING] Found blocking file, exiting now. To load homebridge-knx, remove /home/homebridge-pi/.homebridge/knx-ignore.txt
[2017-02-24 10:16:55] ---
[2017-02-24 10:16:55] Loaded config.json with 0 accessories and 0 platforms.
[2017-02-24 10:16:55] ---
[2017-02-24 10:16:55] Loading 0 platforms...
[2017-02-24 10:16:55] Loading 0 accessories...
Failed to find plugin to handle accessory Desk lamp
```
The failed accessory is a KNX accessory that would have been loaded if homebridge-knx was actually loaded.

@zork17 : you could give that branch a try!
[IRC the installation](https://docs.npmjs.com/cli/install) should look like `sudo npm install -g github:snowdd1/homebridge-knx#ignore-option` *for you global installation approach*

**Comment by zork17** [comment]
@snowdd1 it works perfectly!!!
Thanks.
The 2 instances are working and all other issues that I had because of the 100 limit are solved.

Now I have one last issue with the configuration: hombridge-lifx-lan plugin.
How can I install a plugin locally so it will be visible and loaded only in one homebridge instance. You mentioned before it is possible.
I have 2 homebridge instances. Each one is a service with a different name, that gets a command line with parameters pointing to a different folders where the different config files are.

**Comment by snowdd1** [comment]
For the user account you use to start homebridge (could be "pi", "homebridge" or other) create a new folder in the home directory ("plugins-1"). `cd` into that directory and use npm to install the platform into that directory . It will create a new "node_modules" subfolder. 

For the instance that is supposed to load that platform you add the plugin-1 path by adding the `-P `*path-to-plugin-1* parameter as described here https://github.com/nfarina/homebridge/blob/master/README.md#plugin-development  
To the startup script or service definition. 

Good luck.

**Comment by snowdd1** [comment]
Btw do not use any `sudo ` or `-g` for the npm installation!!!

**Comment by zork17** [comment]
@snowdd1 
I created a folder plugins and run installation:
```
pi@RSBPI:~/plugins $ npm install homebridge-lifx-lan
/home/pi
└─┬ homebridge-lifx-lan@0.2.1
  └─┬ node-lifx@0.8.0
    ├── eventemitter3@2.0.2
    └── lodash@4.17.4

npm WARN enoent ENOENT: no such file or directory, open '/home/pi/package.json'
npm WARN pi No description
npm WARN pi No repository field.
npm WARN pi No README data
npm WARN pi No license field.

```
no "node_modules" subfolder was created
there is no package.json in /home/pi

**Comment by snowdd1** [comment]
Really? Never had that! Please post the result of `ls -ahl ~/plugins `

**Comment by zork17** [comment]
```
pi@RSBPI:~/plugins $ ls -ahl ~/plugins
total 8.0K
drwxrwxrwx  2 pi pi 4.0K Feb 26 17:41 .
drwxr-xr-x 27 pi pi 4.0K Feb 26 17:41 ..
pi@RSBPI:~/plugins $ cd ..
pi@RSBPI:~ $ ls
Desktop     knxd                               knxd-tools_0.10.13-3_armhf.deb    Music         pthsem_2.0.8_armhf.changes  Videos
Documents   knxd_0.10.13-3_armhf.changes       libpthsem20_2.0.8_armhf.deb       node_modules  pthsem_2.0.8.tar.gz
Downloads   knxd_0.10.13-3_armhf.deb           libpthsem-compat_2.0.8_armhf.deb  Pictures      Public
HAP-NodeJS  knxd-dev_0.10.13-3_all.deb         libpthsem-dbg_2.0.8_armhf.deb     plugins       python_games
knx         knxd-examples_0.10.13-3_armhf.deb  libpthsem-dev_2.0.8_armhf.deb     pthsem-2.0.8  Templates
pi@RSBPI:~ $

```
This is very strange. The plugins folder is empty, but there is a node_modules folder next to it - under the home folder, although I ran the install under plugins.
The plugins are working. I used:  -P /home/pi/node_modules

**Comment by imperialjoy** [comment]
Would it be possible to add this ignore option for other plugins too?  The plugin homebridge-mi-aqara has this problem, once its installed, all instances of homebridge must have a mac address and password for the aqara device defined.  It would be best to be able to dynamically load plugins instead of loading everything every time (for example plugin-name-ignore.txt).  But until then, I'd settle for a knx-ignore.txt style solution too, say "aqara-ignore.txt".

**Comment by macjeff** [comment]
I have to look but there is a command to add to your launch command that tells a plug-in directory.   So for each instance I have a separate plugin directory.   Is that what you mean?

Jeff

**Comment by imperialjoy** [comment]
Yes, that works too, I didn't know about -P until you mentioned it.

Its even better as it means only the specified plugins are loaded.  Makes updating a problem though.  An ignore option would leave the plugins in the plugin folder so they could update, but I'll take this for now thanks!
